### ISSUE TITLE
consistency issues

### ISSUE THREAD

**Issue by samoht** [mirage/irmin#1476]
This is a tracking issue for some random consistency issues that users have been seeing with Tezos (for instance [here](https://gitlab.com/tezos/tezos/-/issues/1429)).

The errors are either of the form:

```
Jun  9 14:20:08.739 - validator.block: validation of block BMCKuAMx128DmB8HBtD6cKHnsuVQvhdKbzsnXu9dPhZR4JVNBhc failed
Jun  9 14:20:08.739 - validator.block: Request pushed on 2021-06-09T13:20:07.279-00:00, treated in 122us, completed in 1.459s , 
Jun  9 14:20:08.739 - validator.block: Inconsistent hash:
Jun  9 14:20:08.739 - validator.block:   got: CoV7GccYyD1ZfJFfu6i2aaTLuCxa3NaVGKx9MoNhRhAzEiHLWvjq
Jun  9 14:20:08.739 - validator.block:   expected: CoV89jZC4tnjoSNs6s5ARSRyqTtAeFfrWjk958acUYyeWHDkKxHG
```

After exploring a corrupted store, it appears that a range of hashes are missing from the index. This causes the sequence of operations applied during block validation to the context to diverge (as one `Context.find` would return `false` instead of `true`) and thus the produced context (and its hash) to be different.

Here are some related tasks:
- [x] https://gitlab.com/tezos/tezos/-/merge_requests/3184 should be able to recover a store who faced the problem ; we might need to add a `check-index` command to simply check that the node is healthy
- [x] https://github.com/mirage/irmin/issues/1475 will help by failing early and avoiding the context to diverge
- [x] some broken tests for log_async recovery in case of crash of the merge thread: https://github.com/mirage/index/pull/337
- [x] a possible fix is https://github.com/mirage/index/pull/339

**Comment by Ngoguey42** [comment]
Detailed explanation on how the corrupted store was constructed.

TL;DR: When a merge thread crashes (e.g. with `Out_of_memory`), the program continues.

### Segmentation of the end of the pack file
```
# The `500_001` lost entries
idx:`49_972_463` (x + 0)          at offset `5_254_590_022` is node `CoUkTCmo1ojgvAdtX1oCwy...` (first object with `2 ocurences`)
idx:`50_472_175` (x + 499_712)    at offset `5_313_687_190` is commit `1475916 (n-2)` (last object with `2 ocurences` of 1st half)
idx:`50_472_176` (x + 499_713)    at offset `5_313_687_344` is a blob (first object with `no trace in index`)
idx:`50_472_463` (x + 500_000)    at offset `5_313_717_175` is a node (last object with `no trace in index`)

# The `829` remaining entries of commit `1475917 (n-1)`
idx:`50_472_464` (prev + 1)       at offset `?` (maybe a blob)
idx:`50_472_465` (prev + 1)       at offset `5_313_717_571` is a node (first object with `children by hash (direct children)`)
idx:`50_473_288` (prev + 823)     at offset `5_313_858_947` is a node (last object with `children by hash (direct children)`)
idx:`50_473_292` (prev + 4)       at offset `5_313_859_640` is commit `1475917 (n-1)`

# Reconstruction of the lost entries (499707 entries)
idx:`50_473_293` (prev + 1)       at offset `5_313_859_794` is node `CoUkTCmo1ojgvAdtX1oCwy...` again (first object with `2 ocurences` of 2nd half)
idx:`50_972_999` (prev + 499_706) at offset `5_372_956_378` is commit `1475916 (n-2)` again (last object with `2 ocurences`)

# The content of commit `1475918 (n+0)` (681 entries)
idx:`50_973_680` (prev + 681)     at offset `5_373_043_002` is commit `1475918 (n-0)` and end of pack file
```

### Explanation step by step

1. The tezos-node successfully performed the commit `1475584 (n-334)`, almost filling the `log_async`.
2. While exporting commit `1475585 (n-333)` to disk, the tezos-node was stuck on an index merge that was successful, securing the commit `1475584 (n-334)` out of the `log_async` to the `log` file. The last objects of commit `1475585 (n-333)` were written at the beginning of a fresh `log_async`.
3. Later, the tezos-node successfully performed the commit `1475916 (n-2)`, almost filling the `log_async` (499_713/500_001).
4. The tezos-node successfully constructed the commit `1475917 (n-1)` in memory, started to export to disk, was stuck on merge after inserting 288 entries.
5. The ongoing merge thread crashed with `Out_of_memory`. The next merge started, it kept the `log` (containing commit `1475585 (n-333)`) but erased the `log_async` (https://github.com/mirage/index/blob/79c134da1e4111b618b4040603bb5913f6b0772a/src/index.ml#L824-L833).
6. The main thread continued pushing the remaining objects of commit `1475917 (n-1)`, using `Direct` predecessors for the entries that were suddenly unknown to index.
7. After that successful commit `1475917 (n-1)`, the tezos-node decided to get news of the commits older than `1475917 (n-1)`, seeing that these disappeared, it iterated backward in the past looking for the first existing commit, i.e. `1475584 (n-334)` from `step 1.`. It then started to rebuild everything from there.
8. 499_707 entries were pushed to pack in order to rebuild the commits from `1475585 (n-333)` to `1475916 (n-2)`. (Not all the 500_001 entries were rebuilt here, 288 are explained in `point 9.` and 6 were already pushed during `point 6.`. (`499_707 + 288 + 6 = 500_001`))
9. As commit `1475917 (n-1)` existed in index, it wasn't rebuilt, and the last 288 lost entries were never created again.
10. The tezos-node started working on commit `1475918 (n+0)` and diverged when `Tree.find /big_maps/index/71/14/1f/ee/05/0b/511/key_type` said `None` because `/big_maps/index/71/14` is one of the 288 dangling hashes.

**Comment by samoht** [comment]
Related to that issue, https://github.com/mirage/irmin/pull/1478 allows to check for a store consistency without reconstructing it. @icristescu is working on exposing this to the `tezos-node storage` subcommand.

**Comment by samoht** [comment]
I've made two releases for the index issues:
- index 1.3.2: https://github.com/ocaml/opam-repository/pull/19068 (should be compatible with tezos 9)
- index 1.4.1: https://github.com/ocaml/opam-repository/pull/19069 (should be compatible with tezos 10)

**Comment by samoht** [comment]
Index 1.3.2 and index 1.4.1 are now released in opam-repository.

**Comment by samoht** [comment]
Closing as this is now fixed in index and in tezos' master.
